File: 10.Group_Anagrams/sol.py
import os
import sys
from collections import defaultdict
from typing import List

# ======================================================================
# 🧠 CHALLENGE: Group Anagrams (Python Version)
# ======================================================================
# Description:
# Given an array of strings `strs`, group the anagrams together.
# You can return the answer in any order.
#
# An Anagram is a word or phrase formed by rearranging the letters of a
# different word or phrase, typically using all the original letters exactly once.
#
# 📋 Rules:
# 1. Return a list of lists, where each inner list contains words that are anagrams.
# 2. The order of the groups and the order of words within groups does not matter.
#
# 💡 Examples:
# - group_anagrams(["eat","tea","tan","ate","nat","bat"])
#   => [["bat"],["nat","tan"],["ate","eat","tea"]]
# - group_anagrams([""])  => [[""]]
# - group_anagrams(["a"]) => [["a"]]
# ======================================================================


# region [📚 Reference Solutions] (Solutions hidden as requested)
def group_anagrams1(strs: List[str]) -> List[List[str]]:
    """
    Method: Sorting + Hash Map
    Complexity: Time O(N * K log K) | Space O(N * K)
    (N = number of strings, K = max length of a string)
    """
    # Key: Sorted String (e.g., "aet")
    # Value: List of original strings (e.g., ["eat", "tea", "ate"])
    anagram_map = defaultdict(list)

    for word in strs:
        # 1. Generate the key by sorting characters
        # sorted("eat") becomes ['a', 'e', 't']
        # "".join(...) turns it back into "aet"
        sorted_key = "".join(sorted(word))

        # 2. Append the original word to the correct group
        anagram_map[sorted_key].append(word)

    # 3. Return just the groups (values)
    return list(anagram_map.values())


def group_anagrams2(strs: list[str]) -> list[list[str]]:
    # 建議改名為 groups 或 anagram_map
    groups = defaultdict(list)

    for word in strs:
        # 1. 建立計數陣列 (Frequency Array)
        # 這裡用 List 是為了可以修改內容
        count = [0] * 26

        for char in word:
            count[ord(char) - ord("a")] += 1

        # 2. 關鍵：轉成 Tuple 才能當 Dictionary Key
        # key 像是 (1, 0, 0, 0, 1, ..., 0)
        key = tuple(count)

        groups[key].append(word)

    return list(groups.values())


# endregion


# ======================================================================
#  region [✍️ Practice Area]
#  Please write your solution between the markers below.
# ======================================================================
# <PRACTICE_START>
def group_anagrams(strs):
    # TODO: Implement your solution here.
    return []
# <PRACTICE_END>
# endregion

# ======================================================================
#  region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
# ======================================================================


def reset_practice_area():
    print("\n🔄 Resetting Practice Area to default state...")
    MARKER_START = "# <PRACTICE_" + "START>"
    MARKER_END = "# <PRACTICE_" + "END>"

    default_code_lines = [
        "def group_anagrams(strs):\n",
        "    # TODO: Implement your solution here.\n",
        "    return []\n",
    ]

    try:
        current_file = __file__
        with open(current_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start_idx = -1
        end_idx = -1

        for i, line in enumerate(lines):
            if MARKER_START in line:
                start_idx = i
            if MARKER_END in line:
                end_idx = i

        if start_idx == -1 or end_idx == -1 or start_idx >= end_idx:
            print("⚠️ Error: Markers not found. Reset cancelled.")
            return

        new_content = lines[: start_idx + 1] + default_code_lines + lines[end_idx:]

        with open(current_file, "w", encoding="utf-8") as f:
            f.writelines(new_content)
        print("✨ Reset complete! The file is ready for a fresh start.")
    except Exception as e:
        print(f"⚠️ Reset failed: {e}")


if __name__ == "__main__":
    test_cases = [
        (
            ["eat", "tea", "tan", "ate", "nat", "bat"],
            [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]],
        ),
        ([""], [[""]]),
        (["a"], [["a"]]),
        (
            ["abc", "bca", "cba", "xyz", "zyx", "klm"],
            [["abc", "bca", "cba"], ["xyz", "zyx"], ["klm"]],
        ),
    ]

    print(f"\n🧪 Testing your [group_anagrams] function...\n")

    header = f"{'Input strs':<30} | {'Status':<10}"
    print(header)
    print("-" * 45)

    all_pass = True
    for strs, expected in test_cases:
        try:
            result = group_anagrams(strs)
        except Exception as e:
            result = "Error"

        # Canonicalize results for comparison (Sort inner lists, then sort the outer list)
        is_match = False
        try:
            if isinstance(result, list):
                res_sorted = sorted([sorted(x) for x in result])
                exp_sorted = sorted([sorted(x) for x in expected])
                is_match = res_sorted == exp_sorted
        except:
            is_match = False

        status_icon = "✅ PASS" if is_match else "❌ FAIL"
        if not is_match:
            all_pass = False

        # Format output
        strs_str = str(strs)
        if len(strs_str) > 28:
            strs_str = strs_str[:25] + "..."

        print(f"{strs_str:<30} | {status_icon}")
        if not is_match:
            print(f"   Expected: {expected}")
            print(f"   Actual:   {result}")

    print("-" * 45)
    if all_pass:
        print("\n🎉 Fantastic! All test cases passed.")
        reset_practice_area()
    else:
        print("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")

# endregion

File: 10.Group_Anagrams/sol.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"sort"
	"strings"
)

// ======================================================================
// 🧠 CHALLENGE: Group Anagrams (Go Version)
// ======================================================================
// Description:
// Given an array of strings `strs`, group the anagrams together.
// You can return the answer in any order.
//
// 📋 Rules:
// 1. Return a slice of slices, where each inner slice contains words that are anagrams.
// 2. The order of the groups and the order of words within groups does not matter.
//
// 💡 Examples:
// - GroupAnagrams(["eat","tea","tan","ate","nat","bat"])
//   => [["bat"],["nat","tan"],["ate","eat","tea"]]
// - GroupAnagrams([""])  => [[""]]
// - GroupAnagrams(["a"]) => [["a"]]
// ======================================================================

// #region [📚 Reference Solutions] (Solutions hidden as requested)
// (Focus on implementing your own logic in the Practice Area below!)
// #endregion

// ======================================================================
//
//	#region [✍️ Practice Area]
//	Please write your solution between the markers below.
//
// ======================================================================
// <PRACTICE_START>
func GroupAnagrams(strs []string) [][]string {
	// TODO: Implement your solution here.
	return [][]string{}
}

// <PRACTICE_END>
// #endregion

// ======================================================================
//  #region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
// ======================================================================

func main() {
	runTests()
}

type TestCase struct {
	strs     []string
	expected [][]string
}

func runTests() {
	testCases := []TestCase{
		{
			[]string{"eat", "tea", "tan", "ate", "nat", "bat"},
			[][]string{{"bat"}, {"nat", "tan"}, {"ate", "eat", "tea"}},
		},
		{
			[]string{""},
			[][]string{{""}},
		},
		{
			[]string{"a"},
			[][]string{{"a"}},
		},
	}

	fmt.Printf("\n🧪 Testing your [GroupAnagrams] function...\n\n")

	header := fmt.Sprintf("%-30s | %-10s", "Input strs", "Status")
	fmt.Println(header)
	fmt.Println(strings.Repeat("-", 45))

	allPass := true

	for _, tc := range testCases {
		result := GroupAnagrams(tc.strs)

		// Canonicalize results for comparison
		canonicalize := func(input [][]string) string {
			if len(input) == 0 {
				return "[]"
			}
			// Copy to sort safely
			copySlice := make([][]string, len(input))
			for i, v := range input {
				innerCopy := make([]string, len(v))
				copy(innerCopy, v)
				sort.Strings(innerCopy)
				copySlice[i] = innerCopy
			}

			// Sort the groups themselves for deterministic comparison
			sort.Slice(copySlice, func(i, j int) bool {
				// Sort by the first element of each group (since inner groups are sorted)
				if len(copySlice[i]) > 0 && len(copySlice[j]) > 0 {
					return copySlice[i][0] < copySlice[j][0]
				}
				return len(copySlice[i]) < len(copySlice[j])
			})
			return fmt.Sprintf("%v", copySlice)
		}

		resStr := canonicalize(result)
		expStr := canonicalize(tc.expected)

		isMatch := resStr == expStr
		statusIcon := "✅ PASS"
		if !isMatch {
			statusIcon = "❌ FAIL"
			allPass = false
		}

		// Helper to format string for display
		inputStr := fmt.Sprintf("%v", tc.strs)
		if len(inputStr) > 28 {
			inputStr = inputStr[:25] + "..."
		}

		fmt.Printf("%-30s | %s\n", inputStr, statusIcon)
		if !isMatch {
			// Print details if failed, as simple table might not show enough
			fmt.Printf("   Expected: %v\n", tc.expected)
			fmt.Printf("   Actual:   %v\n", result)
		}
	}

	fmt.Println(strings.Repeat("-", 45))

	if allPass {
		fmt.Println("\n🎉 Fantastic! All test cases passed.")
		resetPracticeArea()
	} else {
		fmt.Println("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
	}
}

func resetPracticeArea() {
	fmt.Println("\n🔄 Resetting Practice Area to default state...")

	markerStart := "// <PRACTICE_" + "START>"
	markerEnd := "// <PRACTICE_" + "END>"

	defaultCode := []string{
		"func GroupAnagrams(strs []string) [][]string {",
		"\t// TODO: Implement your solution here.",
		"\treturn [][]string{}",
		"}",
	}

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		fmt.Println("⚠️ Error: Could not determine file path. Reset cancelled.")
		return
	}

	content, err := os.ReadFile(currentFile)
	if err != nil {
		fmt.Printf("⚠️ Error reading file: %v\n", err)
		return
	}

	lines := strings.Split(string(content), "\n")
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if strings.Contains(line, markerStart) {
			startIdx = i
		} else if strings.Contains(line, markerEnd) {
			endIdx = i
		}
	}

	if startIdx == -1 || endIdx == -1 || startIdx >= endIdx {
		fmt.Println("⚠️ Error: Markers not found or invalid. Reset cancelled.")
		return
	}

	newLines := make([]string, 0)
	newLines = append(newLines, lines[:startIdx+1]...)
	newLines = append(newLines, defaultCode...)
	newLines = append(newLines, lines[endIdx:]...)

	output := strings.Join(newLines, "\n")
	err = os.WriteFile(currentFile, []byte(output), 0644)
	if err != nil {
		fmt.Printf("⚠️ Error writing file: %v\n", err)
		return
	}

	fmt.Println("✨ Reset complete! The file is ready for a fresh start.")
}

// #endregion
